### PR TITLE
Fixes handling bitwise operators for z3 model checker

### LINKED ISSUE #13900: [SMTChecker] ICE due to unimplemented `Z3_OP_BUREM` and `Z3_OP_BSDIV0` when using bitwise `|`, `&` or `^`
## Description
Using the bitwise operator `|`, `&` or `^` triggers an assertion in `Z3Interface::fromZ3Expr()`. Apparently we're not handling some Z3 expression kinds that come up when this code is processed by the SMTChecker.

I checked a few other operators and they do not seem to be affected. We should have a comprehensive test checking them all though.

## Steps to Reproduce
`test.sol`
```solidity
contract C {
    function bitor(uint x, uint y) public pure {
        x | y; // decl_kind is Z3_OP_BUREM
        x & y; // decl_kind is Z3_OP_BUREM
        x ^ y; // decl_kind is Z3_OP_BSDIV0
        assert(true);
    }
}
```
```bash
solc test.sol --model-checker-engine all
```
```
SMT logic error:
/solidity/libsmtutil/Z3Interface.cpp(385): Throw in function solidity::smtutil::Expression solidity::smtutil::Z3Interface::fromZ3Expr(const z3::expr&)
Dynamic exception type: boost::wrapexcept<solidity::smtutil::SMTLogicError>
std::exception::what: SMT assertion failed
[solidity::util::tag_comment*] = SMT assertion failed
```

## Environment

- Compiler version: 0.8.17 and latest `develop`.

### PR DESCRIPTION
**Pull request by pgebal**
Fixes #13900

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -47,6 +47,7 @@ Bugfixes:
  * SMTChecker: Fix internal error when deleting struct member of function type.
  * SMTChecker: Fix internal error when using user-defined types as mapping indices or struct members.
  * SMTChecker: Improved readability for large integers that are powers of two or almost powers of two in error messages.
+ * SMTChecker: Fix compiler error caused by unhandled ``z3`` expressions that come up when bitwise operators are used in solidity code.
```
Comment (leonardoalt) [inline Changelog.md]: Nit:
```suggestion
 * SMTChecker: Fix internal error caused by unhandled ``z3`` expressions that come from the solver when bitwise operators are used.
```

**Comment by leonardoalt** [comment]
This looks good!
There's one odd thing though: in the original issue, the unhandled z3 ast nodes were Z3_OP_BUREM and Z3_OP_BSDIV0, I wonder why you got different ones.
@cameel what's the z3 version you had when you found the issue?
@pgebal maybe it makes sense to add these two as well?

**Comment by cameel** [comment]
> what's the z3 version you had when you found the issue?

4.11.2

File: test/libsolidity/smtCheckerTests/operators/bitwise_operators_do_not_throw_exceptions.sol [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity >=0.0;
```
Comment (cameel) [inline test/libsolidity/smtCheckerTests/operators/bitwise_operators_do_not_throw_exceptions.sol]: We don't include these in tests explicitly. Our test framework includes them automatically.

File: Changelog.md (hunk 1 above)
Comment (cameel) [inline Changelog.md]: This is changelog for 0.8.18, which has already been released. Please move to 0.8.19 section.

File: test/libsolidity/smtCheckerTests/operators/bitwise_operators_do_not_throw_exceptions.sol (hunk 2 above)
Comment (leonardoalt) [reply test/libsolidity/smtCheckerTests/operators/bitwise_operators_do_not_throw_exceptions.sol]: @pgebal these lines above should be removed

File: Changelog.md [hunk 3]
```diff
@@ -47,6 +47,7 @@ Bugfixes:
  * SMTChecker: Fix internal error when deleting struct member of function type.
  * SMTChecker: Fix internal error when using user-defined types as mapping indices or struct members.
  * SMTChecker: Improved readability for large integers that are powers of two or almost powers of two in error messages.
+ * SMTChecker: Fix internal error caused by unhandled ``z3`` expressions that come from the solver when bitwise operators are used.
```
Comment (leonardoalt) [inline Changelog.md]: @pgebal this should be moved to the top of the file under 0.8.19

**Review by leonardoalt — DISMISSED** [review]
lgtm, @cameel ?

**Comment by leonardoalt** [comment]
Actually, can you please squash the commits @pgebal

**Comment by pgebal** [comment]
> Actually, can you please squash the commits @pgebal

done

**Review by leonardoalt — APPROVED** [review]
Looks good. @cameel

**Pull request merged**